File: three-language-version/lessons/16-graphs/graphs.py
"""
Lesson 16: Graphs - Python Examples
Graph representations, traversals, and algorithms
"""

from typing import Dict, List, Set, Optional, Tuple
from collections import defaultdict, deque
import heapq

# =============================================================================
# GRAPH REPRESENTATIONS
# =============================================================================

# Adjacency List - Most common, efficient for sparse graphs
class Graph:
    """Graph using adjacency list representation"""

    def __init__(self, directed: bool = False):
        self.adj: Dict[str, List[str]] = defaultdict(list)
        self.directed = directed

    def add_edge(self, u: str, v: str) -> None:
        self.adj[u].append(v)
        if not self.directed:
            self.adj[v].append(u)
        # Ensure both nodes exist even if isolated
        if v not in self.adj:
            self.adj[v] = []

    def neighbors(self, node: str) -> List[str]:
        return self.adj[node]

    def nodes(self) -> List[str]:
        return list(self.adj.keys())

    def __repr__(self) -> str:
        return "\n".join(f"{k}: {v}" for k, v in self.adj.items())


print("=== Adjacency List Representation ===")
g = Graph()
edges = [("A", "B"), ("A", "C"), ("B", "D"), ("C", "D"), ("D", "E")]
for u, v in edges:
    g.add_edge(u, v)
print(g)


# Adjacency Matrix - Good for dense graphs, O(1) edge lookup
class GraphMatrix:
    """Graph using adjacency matrix"""

    def __init__(self, nodes: List[str]):
        self.nodes = nodes
        self.node_idx = {node: i for i, node in enumerate(nodes)}
        n = len(nodes)
        self.matrix = [[0] * n for _ in range(n)]

    def add_edge(self, u: str, v: str, weight: int = 1) -> None:
        i, j = self.node_idx[u], self.node_idx[v]
        self.matrix[i][j] = weight
        self.matrix[j][i] = weight  # Undirected

    def has_edge(self, u: str, v: str) -> bool:
        i, j = self.node_idx[u], self.node_idx[v]
        return self.matrix[i][j] != 0

    def print_matrix(self) -> None:
        print("   " + " ".join(self.nodes))
        for i, node in enumerate(self.nodes):
            print(f"{node}: {self.matrix[i]}")


print("\n=== Adjacency Matrix Representation ===")
gm = GraphMatrix(["A", "B", "C", "D"])
gm.add_edge("A", "B")
gm.add_edge("A", "C")
gm.add_edge("B", "D")
gm.print_matrix()


# =============================================================================
# DEPTH-FIRST SEARCH (DFS)
# =============================================================================

def dfs_recursive(graph: Graph, start: str, visited: Set[str] = None) -> List[str]:
    """DFS using recursion"""
    if visited is None:
        visited = set()

    visited.add(start)
    result = [start]

    for neighbor in graph.neighbors(start):
        if neighbor not in visited:
            result.extend(dfs_recursive(graph, neighbor, visited))

    return result


def dfs_iterative(graph: Graph, start: str) -> List[str]:
    """DFS using explicit stack"""
    visited = set()
    stack = [start]
    result = []

    while stack:
        node = stack.pop()
        if node not in visited:
            visited.add(node)
            result.append(node)
            # Add neighbors in reverse for consistent order
            for neighbor in reversed(graph.neighbors(node)):
                if neighbor not in visited:
                    stack.append(neighbor)

    return result


print("\n=== Depth-First Search (DFS) ===")
print(f"DFS recursive from A: {dfs_recursive(g, 'A')}")
print(f"DFS iterative from A: {dfs_iterative(g, 'A')}")


# =============================================================================
# BREADTH-FIRST SEARCH (BFS)
# =============================================================================

def bfs(graph: Graph, start: str) -> List[str]:
    """BFS using queue - explores level by level"""
    visited = {start}
    queue = deque([start])
    result = []

    while queue:
        node = queue.popleft()
        result.append(node)

        for neighbor in graph.neighbors(node):
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append(neighbor)

    return result


def bfs_with_levels(graph: Graph, start: str) -> List[List[str]]:
    """BFS returning nodes by level/distance"""
    visited = {start}
    queue = deque([start])
    levels = []

    while queue:
        level = []
        for _ in range(len(queue)):
            node = queue.popleft()
            level.append(node)

            for neighbor in graph.neighbors(node):
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)

        levels.append(level)

    return levels


print("\n=== Breadth-First Search (BFS) ===")
print(f"BFS from A: {bfs(g, 'A')}")
print(f"BFS levels from A: {bfs_with_levels(g, 'A')}")


# =============================================================================
# SHORTEST PATH (Unweighted)
# =============================================================================

def shortest_path_bfs(graph: Graph, start: str, end: str) -> Optional[List[str]]:
    """Find shortest path using BFS (unweighted)"""
    if start == end:
        return [start]

    visited = {start}
    queue = deque([(start, [start])])

    while queue:
        node, path = queue.popleft()

        for neighbor in graph.neighbors(node):
            if neighbor not in visited:
                new_path = path + [neighbor]
                if neighbor == end:
                    return new_path
                visited.add(neighbor)
                queue.append((neighbor, new_path))

    return None


print("\n=== Shortest Path (Unweighted) ===")
print(f"Shortest path A to E: {shortest_path_bfs(g, 'A', 'E')}")


# =============================================================================
# WEIGHTED GRAPH
# =============================================================================

class WeightedGraph:
    """Graph with weighted edges"""

    def __init__(self, directed: bool = False):
        self.adj: Dict[str, List[Tuple[str, int]]] = defaultdict(list)
        self.directed = directed

    def add_edge(self, u: str, v: str, weight: int) -> None:
        self.adj[u].append((v, weight))
        if not self.directed:
            self.adj[v].append((u, weight))

    def neighbors(self, node: str) -> List[Tuple[str, int]]:
        return self.adj[node]

    def nodes(self) -> List[str]:
        return list(self.adj.keys())


# =============================================================================
# DIJKSTRA'S ALGORITHM
# =============================================================================

def dijkstra(graph: WeightedGraph, start: str) -> Dict[str, int]:
    """
    Dijkstra's algorithm for shortest paths
    Returns distances from start to all reachable nodes
    O((V + E) log V) with binary heap
    """
    distances = {node: float('inf') for node in graph.nodes()}
    distances[start] = 0
    pq = [(0, start)]  # (distance, node)
    visited = set()

    while pq:
        dist, node = heapq.heappop(pq)

        if node in visited:
            continue
        visited.add(node)

        for neighbor, weight in graph.neighbors(node):
            new_dist = dist + weight
            if new_dist < distances[neighbor]:
                distances[neighbor] = new_dist
                heapq.heappush(pq, (new_dist, neighbor))

    return distances


def dijkstra_with_path(graph: WeightedGraph, start: str, end: str) -> Tuple[int, List[str]]:
    """Dijkstra's that also returns the path"""
    distances = {node: float('inf') for node in graph.nodes()}
    distances[start] = 0
    previous = {}
    pq = [(0, start)]
    visited = set()

    while pq:
        dist, node = heapq.heappop(pq)

        if node in visited:
            continue
        visited.add(node)

        if node == end:
            break

        for neighbor, weight in graph.neighbors(node):
            new_dist = dist + weight
            if new_dist < distances[neighbor]:
                distances[neighbor] = new_dist
                previous[neighbor] = node
                heapq.heappush(pq, (new_dist, neighbor))

    # Reconstruct path
    path = []
    current = end
    while current in previous:
        path.append(current)
        current = previous[current]
    path.append(start)
    path.reverse()

    return distances[end], path


print("\n=== Dijkstra's Algorithm ===")
wg = WeightedGraph()
wg.add_edge("A", "B", 4)
wg.add_edge("A", "C", 2)
wg.add_edge("B", "C", 1)
wg.add_edge("B", "D", 5)
wg.add_edge("C", "D", 8)
wg.add_edge("C", "E", 10)
wg.add_edge("D", "E", 2)

distances = dijkstra(wg, "A")
print(f"Distances from A: {distances}")

dist, path = dijkstra_with_path(wg, "A", "E")
print(f"Shortest path A to E: {path} (distance: {dist})")


# =============================================================================
# CYCLE DETECTION
# =============================================================================

def has_cycle_undirected(graph: Graph) -> bool:
    """Detect cycle in undirected graph using DFS"""
    visited = set()

    def dfs(node: str, parent: Optional[str]) -> bool:
        visited.add(node)
        for neighbor in graph.neighbors(node):
            if neighbor not in visited:
                if dfs(neighbor, node):
                    return True
            elif neighbor != parent:
                return True
        return False

    for node in graph.nodes():
        if node not in visited:
            if dfs(node, None):
                return True
    return False


def has_cycle_directed(graph: Graph) -> bool:
    """Detect cycle in directed graph using colors"""
    WHITE, GRAY, BLACK = 0, 1, 2
    colors = {node: WHITE for node in graph.nodes()}

    def dfs(node: str) -> bool:
        colors[node] = GRAY
        for neighbor in graph.neighbors(node):
            if colors[neighbor] == GRAY:
                return True  # Back edge = cycle
            if colors[neighbor] == WHITE and dfs(neighbor):
                return True
        colors[node] = BLACK
        return False

    for node in graph.nodes():
        if colors[node] == WHITE:
            if dfs(node):
                return True
    return False


print("\n=== Cycle Detection ===")
cycle_graph = Graph()
cycle_graph.add_edge("A", "B")
cycle_graph.add_edge("B", "C")
cycle_graph.add_edge("C", "A")  # Creates cycle
print(f"Graph with cycle: {has_cycle_undirected(cycle_graph)}")


# =============================================================================
# TOPOLOGICAL SORT (DAG)
# =============================================================================

def topological_sort(graph: Graph) -> List[str]:
    """
    Topological sort using DFS
    Only valid for DAGs (Directed Acyclic Graphs)
    """
    visited = set()
    result = []

    def dfs(node: str):
        visited.add(node)
        for neighbor in graph.neighbors(node):
            if neighbor not in visited:
                dfs(neighbor)
        result.append(node)

    for node in graph.nodes():
        if node not in visited:
            dfs(node)

    result.reverse()
    return result


def topological_sort_kahn(graph: Graph) -> List[str]:
    """Topological sort using Kahn's algorithm (BFS)"""
    # Calculate in-degrees
    in_degree = defaultdict(int)
    for node in graph.nodes():
        for neighbor in graph.neighbors(node):
            in_degree[neighbor] += 1

    # Start with nodes that have no incoming edges
    queue = deque([node for node in graph.nodes() if in_degree[node] == 0])
    result = []

    while queue:
        node = queue.popleft()
        result.append(node)
        for neighbor in graph.neighbors(node):
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    return result if len(result) == len(graph.nodes()) else []  # Empty if cycle


print("\n=== Topological Sort ===")
dag = Graph(directed=True)
dag.add_edge("A", "C")
dag.add_edge("B", "C")
dag.add_edge("B", "D")
dag.add_edge("C", "E")
dag.add_edge("D", "F")
dag.add_edge("E", "F")
print(f"Topological order: {topological_sort(dag)}")


# =============================================================================
# CONNECTED COMPONENTS
# =============================================================================

def connected_components(graph: Graph) -> List[List[str]]:
    """Find all connected components in undirected graph"""
    visited = set()
    components = []

    def dfs(node: str, component: List[str]):
        visited.add(node)
        component.append(node)
        for neighbor in graph.neighbors(node):
            if neighbor not in visited:
                dfs(neighbor, component)

    for node in graph.nodes():
        if node not in visited:
            component = []
            dfs(node, component)
            components.append(component)

    return components


print("\n=== Connected Components ===")
disconnected = Graph()
disconnected.add_edge("A", "B")
disconnected.add_edge("B", "C")
disconnected.add_edge("D", "E")  # Separate component
disconnected.add_edge("F", "F")  # Isolated (self-loop for node existence)
# Add isolated node G
disconnected.adj["G"] = []
print(f"Components: {connected_components(disconnected)}")


# =============================================================================
# MINIMUM SPANNING TREE (PRIM'S)
# =============================================================================

def prim_mst(graph: WeightedGraph) -> List[Tuple[str, str, int]]:
    """
    Prim's algorithm for Minimum Spanning Tree
    Returns list of edges in MST
    """
    if not graph.nodes():
        return []

    start = graph.nodes()[0]
    visited = {start}
    edges = []
    pq = [(weight, start, neighbor) for neighbor, weight in graph.neighbors(start)]
    heapq.heapify(pq)

    while pq and len(visited) < len(graph.nodes()):
        weight, u, v = heapq.heappop(pq)
        if v in visited:
            continue

        visited.add(v)
        edges.append((u, v, weight))

        for neighbor, w in graph.neighbors(v):
            if neighbor not in visited:
                heapq.heappush(pq, (w, v, neighbor))

    return edges


print("\n=== Minimum Spanning Tree (Prim's) ===")
mst_edges = prim_mst(wg)
total_weight = sum(e[2] for e in mst_edges)
print(f"MST edges: {mst_edges}")
print(f"Total weight: {total_weight}")


# =============================================================================
# COMPARISON
# =============================================================================

print("\n=== Graph Algorithm Summary ===")
print("""
Traversals:
  - DFS: O(V + E) - uses stack/recursion
  - BFS: O(V + E) - uses queue, finds shortest path (unweighted)

Shortest Path:
  - BFS: O(V + E) - unweighted graphs
  - Dijkstra: O((V + E) log V) - weighted graphs, non-negative weights
  - Bellman-Ford: O(V * E) - handles negative weights

Minimum Spanning Tree:
  - Prim's: O((V + E) log V) - good for dense graphs
  - Kruskal's: O(E log E) - good for sparse graphs

Topological Sort:
  - DFS: O(V + E) - for DAGs only
  - Kahn's (BFS): O(V + E) - also detects cycles

Connected Components:
  - DFS/BFS: O(V + E)

Representation:
  - Adjacency List: O(V + E) space, good for sparse
  - Adjacency Matrix: O(V²) space, O(1) edge lookup
""")

File: three-language-version/lessons/16-graphs/graphs.cpp
/**
 * Lesson 16: Graphs - C++ Examples
 * Graph representations, traversals, and algorithms
 */

#include <iostream>
#include <vector>
#include <queue>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <set>
#include <algorithm>
#include <limits>
#include <functional>

using namespace std;

// =============================================================================
// ADJACENCY LIST GRAPH
// =============================================================================

class Graph {
public:
    unordered_map<string, vector<string>> adj;
    bool directed;

    Graph(bool dir = false) : directed(dir) {}

    void addEdge(const string& u, const string& v) {
        adj[u].push_back(v);
        if (!directed) {
            adj[v].push_back(u);
        }
        // Ensure both nodes exist
        if (adj.find(v) == adj.end()) {
            adj[v] = {};
        }
    }

    vector<string> neighbors(const string& node) const {
        auto it = adj.find(node);
        return it != adj.end() ? it->second : vector<string>{};
    }

    vector<string> nodes() const {
        vector<string> result;
        for (const auto& p : adj) {
            result.push_back(p.first);
        }
        return result;
    }

    void print() const {
        for (const auto& p : adj) {
            cout << p.first << ": ";
            for (const auto& n : p.second) {
                cout << n << " ";
            }
            cout << endl;
        }
    }
};

void graph_representation_demo() {
    cout << "=== Adjacency List Representation ===" << endl;

    Graph g;
    g.addEdge("A", "B");
    g.addEdge("A", "C");
    g.addEdge("B", "D");
    g.addEdge("C", "D");
    g.addEdge("D", "E");

    g.print();
}

// =============================================================================
// DEPTH-FIRST SEARCH (DFS)
// =============================================================================

void dfsRecursive(const Graph& g, const string& node,
                  unordered_set<string>& visited, vector<string>& result) {
    visited.insert(node);
    result.push_back(node);

    for (const auto& neighbor : g.neighbors(node)) {
        if (visited.find(neighbor) == visited.end()) {
            dfsRecursive(g, neighbor, visited, result);
        }
    }
}

vector<string> dfs(const Graph& g, const string& start) {
    unordered_set<string> visited;
    vector<string> result;
    dfsRecursive(g, start, visited, result);
    return result;
}

vector<string> dfsIterative(const Graph& g, const string& start) {
    unordered_set<string> visited;
    vector<string> result;
    stack<string> stk;
    stk.push(start);

    while (!stk.empty()) {
        string node = stk.top();
        stk.pop();

        if (visited.find(node) != visited.end()) continue;
        visited.insert(node);
        result.push_back(node);

        auto neighbors = g.neighbors(node);
        for (auto it = neighbors.rbegin(); it != neighbors.rend(); ++it) {
            if (visited.find(*it) == visited.end()) {
                stk.push(*it);
            }
        }
    }
    return result;
}

void dfs_demo() {
    cout << "\n=== Depth-First Search (DFS) ===" << endl;

    Graph g;
    g.addEdge("A", "B");
    g.addEdge("A", "C");
    g.addEdge("B", "D");
    g.addEdge("C", "D");
    g.addEdge("D", "E");

    cout << "DFS recursive from A: ";
    for (const auto& node : dfs(g, "A")) cout << node << " ";
    cout << endl;

    cout << "DFS iterative from A: ";
    for (const auto& node : dfsIterative(g, "A")) cout << node << " ";
    cout << endl;
}

// =============================================================================
// BREADTH-FIRST SEARCH (BFS)
// =============================================================================

vector<string> bfs(const Graph& g, const string& start) {
    unordered_set<string> visited;
    vector<string> result;
    queue<string> q;

    visited.insert(start);
    q.push(start);

    while (!q.empty()) {
        string node = q.front();
        q.pop();
        result.push_back(node);

        for (const auto& neighbor : g.neighbors(node)) {
            if (visited.find(neighbor) == visited.end()) {
                visited.insert(neighbor);
                q.push(neighbor);
            }
        }
    }
    return result;
}

vector<string> shortestPathBFS(const Graph& g, const string& start, const string& end) {
    if (start == end) return {start};

    unordered_set<string> visited;
    queue<pair<string, vector<string>>> q;

    visited.insert(start);
    q.push({start, {start}});

    while (!q.empty()) {
        auto [node, path] = q.front();
        q.pop();

        for (const auto& neighbor : g.neighbors(node)) {
            if (visited.find(neighbor) == visited.end()) {
                vector<string> newPath = path;
                newPath.push_back(neighbor);

                if (neighbor == end) {
                    return newPath;
                }

                visited.insert(neighbor);
                q.push({neighbor, newPath});
            }
        }
    }
    return {};  // No path found
}

void bfs_demo() {
    cout << "\n=== Breadth-First Search (BFS) ===" << endl;

    Graph g;
    g.addEdge("A", "B");
    g.addEdge("A", "C");
    g.addEdge("B", "D");
    g.addEdge("C", "D");
    g.addEdge("D", "E");

    cout << "BFS from A: ";
    for (const auto& node : bfs(g, "A")) cout << node << " ";
    cout << endl;

    cout << "Shortest path A to E: ";
    for (const auto& node : shortestPathBFS(g, "A", "E")) cout << node << " ";
    cout << endl;
}

// =============================================================================
// WEIGHTED GRAPH
// =============================================================================

class WeightedGraph {
public:
    unordered_map<string, vector<pair<string, int>>> adj;

    void addEdge(const string& u, const string& v, int weight) {
        adj[u].push_back({v, weight});
        adj[v].push_back({u, weight});  // Undirected
    }

    vector<pair<string, int>> neighbors(const string& node) const {
        auto it = adj.find(node);
        return it != adj.end() ? it->second : vector<pair<string, int>>{};
    }

    vector<string> nodes() const {
        vector<string> result;
        for (const auto& p : adj) {
            result.push_back(p.first);
        }
        return result;
    }
};

// =============================================================================
// DIJKSTRA'S ALGORITHM
// =============================================================================

unordered_map<string, int> dijkstra(const WeightedGraph& g, const string& start) {
    unordered_map<string, int> distances;
    for (const auto& node : g.nodes()) {
        distances[node] = numeric_limits<int>::max();
    }
    distances[start] = 0;

    // Min-heap: (distance, node)
    priority_queue<pair<int, string>, vector<pair<int, string>>, greater<>> pq;
    pq.push({0, start});

    unordered_set<string> visited;

    while (!pq.empty()) {
        auto [dist, node] = pq.top();
        pq.pop();

        if (visited.count(node)) continue;
        visited.insert(node);

        for (const auto& [neighbor, weight] : g.neighbors(node)) {
            int newDist = dist + weight;
            if (newDist < distances[neighbor]) {
                distances[neighbor] = newDist;
                pq.push({newDist, neighbor});
            }
        }
    }

    return distances;
}

pair<int, vector<string>> dijkstraWithPath(const WeightedGraph& g,
                                           const string& start, const string& end) {
    unordered_map<string, int> distances;
    unordered_map<string, string> previous;

    for (const auto& node : g.nodes()) {
        distances[node] = numeric_limits<int>::max();
    }
    distances[start] = 0;

    priority_queue<pair<int, string>, vector<pair<int, string>>, greater<>> pq;
    pq.push({0, start});
    unordered_set<string> visited;

    while (!pq.empty()) {
        auto [dist, node] = pq.top();
        pq.pop();

        if (visited.count(node)) continue;
        visited.insert(node);

        if (node == end) break;

        for (const auto& [neighbor, weight] : g.neighbors(node)) {
            int newDist = dist + weight;
            if (newDist < distances[neighbor]) {
                distances[neighbor] = newDist;
                previous[neighbor] = node;
                pq.push({newDist, neighbor});
            }
        }
    }

    // Reconstruct path
    vector<string> path;
    string current = end;
    while (previous.count(current)) {
        path.push_back(current);
        current = previous[current];
    }
    path.push_back(start);
    reverse(path.begin(), path.end());

    return {distances[end], path};
}

void dijkstra_demo() {
    cout << "\n=== Dijkstra's Algorithm ===" << endl;

    WeightedGraph wg;
    wg.addEdge("A", "B", 4);
    wg.addEdge("A", "C", 2);
    wg.addEdge("B", "C", 1);
    wg.addEdge("B", "D", 5);
    wg.addEdge("C", "D", 8);
    wg.addEdge("C", "E", 10);
    wg.addEdge("D", "E", 2);

    auto distances = dijkstra(wg, "A");
    cout << "Distances from A: ";
    for (const auto& [node, dist] : distances) {
        cout << node << "=" << dist << " ";
    }
    cout << endl;

    auto [dist, path] = dijkstraWithPath(wg, "A", "E");
    cout << "Shortest path A to E: ";
    for (const auto& node : path) cout << node << " ";
    cout << "(distance: " << dist << ")" << endl;
}

// =============================================================================
// CYCLE DETECTION
// =============================================================================

bool hasCycleUndirectedDFS(const Graph& g, const string& node,
                           const string& parent, unordered_set<string>& visited) {
    visited.insert(node);
    for (const auto& neighbor : g.neighbors(node)) {
        if (visited.find(neighbor) == visited.end()) {
            if (hasCycleUndirectedDFS(g, neighbor, node, visited)) {
                return true;
            }
        } else if (neighbor != parent) {
            return true;
        }
    }
    return false;
}

bool hasCycleUndirected(const Graph& g) {
    unordered_set<string> visited;
    for (const auto& node : g.nodes()) {
        if (visited.find(node) == visited.end()) {
            if (hasCycleUndirectedDFS(g, node, "", visited)) {
                return true;
            }
        }
    }
    return false;
}

void cycle_demo() {
    cout << "\n=== Cycle Detection ===" << endl;

    Graph cycleGraph;
    cycleGraph.addEdge("A", "B");
    cycleGraph.addEdge("B", "C");
    cycleGraph.addEdge("C", "A");  // Creates cycle

    cout << "Graph has cycle: " << boolalpha << hasCycleUndirected(cycleGraph) << endl;

    Graph acyclic;
    acyclic.addEdge("A", "B");
    acyclic.addEdge("B", "C");
    cout << "Acyclic graph has cycle: " << hasCycleUndirected(acyclic) << endl;
}

// =============================================================================
// TOPOLOGICAL SORT
// =============================================================================

void topSortDFS(const Graph& g, const string& node,
                unordered_set<string>& visited, vector<string>& result) {
    visited.insert(node);
    for (const auto& neighbor : g.neighbors(node)) {
        if (visited.find(neighbor) == visited.end()) {
            topSortDFS(g, neighbor, visited, result);
        }
    }
    result.push_back(node);
}

vector<string> topologicalSort(const Graph& g) {
    unordered_set<string> visited;
    vector<string> result;

    for (const auto& node : g.nodes()) {
        if (visited.find(node) == visited.end()) {
            topSortDFS(g, node, visited, result);
        }
    }

    reverse(result.begin(), result.end());
    return result;
}

void topsort_demo() {
    cout << "\n=== Topological Sort ===" << endl;

    Graph dag(true);  // Directed
    dag.addEdge("A", "C");
    dag.addEdge("B", "C");
    dag.addEdge("B", "D");
    dag.addEdge("C", "E");
    dag.addEdge("D", "F");
    dag.addEdge("E", "F");

    cout << "Topological order: ";
    for (const auto& node : topologicalSort(dag)) cout << node << " ";
    cout << endl;
}

// =============================================================================
// CONNECTED COMPONENTS
// =============================================================================

void componentDFS(const Graph& g, const string& node,
                  unordered_set<string>& visited, vector<string>& component) {
    visited.insert(node);
    component.push_back(node);
    for (const auto& neighbor : g.neighbors(node)) {
        if (visited.find(neighbor) == visited.end()) {
            componentDFS(g, neighbor, visited, component);
        }
    }
}

vector<vector<string>> connectedComponents(const Graph& g) {
    unordered_set<string> visited;
    vector<vector<string>> components;

    for (const auto& node : g.nodes()) {
        if (visited.find(node) == visited.end()) {
            vector<string> component;
            componentDFS(g, node, visited, component);
            components.push_back(component);
        }
    }

    return components;
}

void components_demo() {
    cout << "\n=== Connected Components ===" << endl;

    Graph disconnected;
    disconnected.addEdge("A", "B");
    disconnected.addEdge("B", "C");
    disconnected.addEdge("D", "E");
    disconnected.adj["F"] = {};  // Isolated node

    auto components = connectedComponents(disconnected);
    cout << "Components: ";
    for (const auto& comp : components) {
        cout << "[";
        for (const auto& node : comp) cout << node << " ";
        cout << "] ";
    }
    cout << endl;
}

// =============================================================================
// MINIMUM SPANNING TREE (Prim's)
// =============================================================================

vector<tuple<string, string, int>> primMST(const WeightedGraph& g) {
    if (g.nodes().empty()) return {};

    string start = g.nodes()[0];
    unordered_set<string> visited;
    visited.insert(start);

    vector<tuple<string, string, int>> mst;

    // Min-heap: (weight, from, to)
    priority_queue<tuple<int, string, string>,
                   vector<tuple<int, string, string>>,
                   greater<>> pq;

    for (const auto& [neighbor, weight] : g.neighbors(start)) {
        pq.push({weight, start, neighbor});
    }

    while (!pq.empty() && visited.size() < g.nodes().size()) {
        auto [weight, u, v] = pq.top();
        pq.pop();

        if (visited.count(v)) continue;

        visited.insert(v);
        mst.push_back({u, v, weight});

        for (const auto& [neighbor, w] : g.neighbors(v)) {
            if (!visited.count(neighbor)) {
                pq.push({w, v, neighbor});
            }
        }
    }

    return mst;
}

void mst_demo() {
    cout << "\n=== Minimum Spanning Tree (Prim's) ===" << endl;

    WeightedGraph wg;
    wg.addEdge("A", "B", 4);
    wg.addEdge("A", "C", 2);
    wg.addEdge("B", "C", 1);
    wg.addEdge("B", "D", 5);
    wg.addEdge("C", "D", 8);
    wg.addEdge("D", "E", 2);

    auto mst = primMST(wg);
    int totalWeight = 0;
    cout << "MST edges: ";
    for (const auto& [u, v, w] : mst) {
        cout << "(" << u << "-" << v << ":" << w << ") ";
        totalWeight += w;
    }
    cout << "\nTotal weight: " << totalWeight << endl;
}

// =============================================================================
// MAIN
// =============================================================================

int main() {
    graph_representation_demo();
    dfs_demo();
    bfs_demo();
    dijkstra_demo();
    cycle_demo();
    topsort_demo();
    components_demo();
    mst_demo();

    cout << "\n=== Graph Algorithm Summary ===" << endl;
    cout << R"(
Traversals:
  - DFS: O(V + E) - stack/recursion
  - BFS: O(V + E) - queue, shortest path (unweighted)

Shortest Path:
  - BFS: O(V + E) - unweighted
  - Dijkstra: O((V + E) log V) - weighted, non-negative

MST:
  - Prim's: O((V + E) log V)
  - Kruskal's: O(E log E)

Topological Sort: O(V + E) - DAGs only
Connected Components: O(V + E)
)" << endl;

    return 0;
}
